torchao/prototype/quantized_training/int8.py

```diff
@@ -0,0 +1,271 @@
+    """INT8 symmetric quantization weight, with absmax scaling [-127, 127]. The main difference
+    of this tensor subclass from AffineQuantizedTensor:
+    1. `F.linear` is differentiable i.e. backward is defined.
+    2. All in-place ops, such as `aten.copy_`, will perform stochastic rounding.
+        `Int8QTLinearWeight.from_float()` does not perform stochastic rounding.
+    3. The numerics for quantization is slightly different. See `Int8QTLinearWeight.quantize()`
+        for more details.
+    """
+
+    implements = classmethod(_implements)
+    __torch_function__ = classmethod(_dispatch__torch_function__)
+    __torch_dispatch__ = classmethod(_dispatch__torch_dispatch__)
+
+    @staticmethod
+    @torch._dynamo.disable
+    def __new__(cls, int_data: Tensor, scale: Tensor):
+        return Tensor._make_wrapper_subclass(
+            cls,
+            int_data.shape,
+            dtype=scale.dtype,
+            device=int_data.device,
+        )
+
+    @torch._dynamo.disable
+    def __init__(self, int_data: Tensor, scale: Tensor):
+        """Create a symmetric quantized INT8 weight. This tensor will appear to have the same dtype
+        as `scale.dtype`. All in-place update ops will perform stochastic rounding.
+        """
+        # NOTE: should scale always be FP32?
+        assert int_data.dtype is torch.int8
+        assert int_data.ndim == 2
+        assert scale.ndim == 1
+        self.int_data = int_data
+        self.scale = scale
+
+    def __tensor_flatten__(self):
+        return ["int_data", "scale"], []
+
+    @classmethod
+    def __tensor_unflatten__(cls, tensor_data_dict, tensor_attributes, outer_size=None, outer_stride=None):
+        return cls(tensor_data_dict["int_data"], tensor_data_dict["scale"], *tensor_attributes)
+
+    @staticmethod
+    @torch.no_grad()
+    def quantize(tensor: Tensor, stochastic_rounding: bool = False):
+        """Normal rounding will always round down small changes in weight update. To tackle this problem,
+        stochastic rounding can be used, which has a low chance, but not zero, of rounding up. The
+        probability of rounding up is equal to x - ⌊x⌋, which indicates how close the value is to the next
+        integer value. Thus, stochastic rounding also approximates the floating point value exactly.
+
+        Currently this function differs from AQT's `int8_weight_only()` in the following way:
+        1. Precision: AQT keeps original dtype when doing quantization, while this function upcasts input
+        to FP32 before quantization, and downcast scale to original dtype.
+        2. Calculate scale: AQT uses `input.abs().amax() / 127.5`, while `input.abs().amax() / 127` is
+        done here.
+        3. Apply scale: AQT uses `input * (1 / scale)`, while this function performs `input / scale`.
+        """
+        original_dtype = tensor.dtype
+        tensor = tensor.float()
+
+        # absmax symmetric quantization
+        scale = tensor.abs().amax(-1) / 127
```

Comment:
so does this mean the quant_min/quant_max for int8 will be (-127, 127)? we could match this in post training quant by explicitly setting quant_min/quant_max I think